File: source/go/dynamic_array.go
package main

import "fmt"

// Dynamic array that grows by doubling
var capacity = 4
var size = 0
var data []int

func grow() {
	newCap := capacity * 2
	newData := make([]int, newCap)
	copy(newData, data)
	data = newData
	capacity = newCap
}

func push(val int) {
	if size == capacity {
		grow()
	}
	data[size] = val
	size++
}

func printAll() {
	fmt.Print("values:")
	for i := 0; i < size; i++ {
		fmt.Printf(" %d", data[i])
	}
	fmt.Println()
}

func main() {
	data = make([]int, capacity)
	for i := 1; i <= 10; i++ {
		push(i)
	}
	fmt.Printf("capacity: %d\n", capacity)
	printAll()
}

File: source/javascript/dynamic_array.js
// Dynamic array that grows (simulating alloc/realloc pattern)
let capacity = 4;
let size = 0;
let data = new Int32Array(capacity);

function grow() {
    const newCap = capacity * 2;
    const newData = new Int32Array(newCap);
    newData.set(data.subarray(0, size));
    data = newData;
    capacity = newCap;
}

function push(val) {
    if (size === capacity) grow();
    data[size++] = val;
}

function printAll() {
    process.stdout.write("values:");
    for (let i = 0; i < size; i++) process.stdout.write(` ${data[i]}`);
    process.stdout.write("\n");
}

for (let i = 1; i <= 10; i++) push(i);
console.log(`capacity: ${capacity}`);
printAll();
